{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pysam\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import collections\n",
    "import seaborn as sns\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "name = 'timp3_3UTR'\n",
    "chrom,start,end = 'chr10',86345383,86350038\n",
    "\n",
    "wiggle_header_stem = 'track type=wiggle_0 name={} visibility=full maxHeightPixels=100:50:20 graphType=point color={},{},{}\\nfixedStep chrom={} start={} step=1\\n'\n",
    "\n",
    "\n",
    "ont_palettes = {\n",
    "    'L6_IT':sns.color_palette('Blues'),\n",
    "    'L23_IT':sns.color_palette('Oranges'),\n",
    "    'Astro':sns.color_palette('Purples'),\n",
    "    'Sst':sns.color_palette('Greens'),\n",
    "}\n",
    "\n",
    "for ont,palette in ont_palettes.items():\n",
    "    stem = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/*{}.bam'.format(ont)\n",
    "    bam_paths = sorted(glob.glob(stem))\n",
    "    \n",
    "    for i,bam_path in enumerate(bam_paths):\n",
    "        color = palette[i]\n",
    "        r,g,b = [int(255*c) for c in color]\n",
    "\n",
    "        with pysam.AlignmentFile(bam_path) as bam:\n",
    "            sample_name = os.path.basename(bam_path).split('.')[0]+'_CDF'\n",
    "            wiggle_header = wiggle_header_stem.format(\n",
    "                sample_name,\n",
    "                r,g,b,\n",
    "                chrom,start,end-start+1,\n",
    "            )\n",
    "            wig_path = bam_path.replace('.bam','_{}.wig'.format(name))\n",
    "            with open(wig_path,'w') as wig:\n",
    "                wig.write(wiggle_header)\n",
    "                base_counts = {pos:0 for pos in range(start,end+1)}\n",
    "                for r in bam.fetch(chrom,start,end):\n",
    "                    if r.pos not in base_counts:\n",
    "                        continue\n",
    "                    base_counts[r.pos] += 1\n",
    "\n",
    "                cum_counts = np.cumsum(list(base_counts.values()))\n",
    "                wig.write('\\n'.join(str(c) for c in cum_counts))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mouse1_Endo_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_Endo_Flt1_3UTR.wig\n",
      "Mouse2_Endo_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse2_Endo_Flt1_3UTR.wig\n",
      "Mouse3_Endo_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_Endo_Flt1_3UTR.wig\n",
      "Mouse4_Endo_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse4_Endo_Flt1_3UTR.wig\n",
      "ss2_Endo_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/ss2_Endo_Flt1_3UTR.wig\n",
      "Mouse1_L5_IT_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_L5_IT_Flt1_3UTR.wig\n",
      "Mouse2_L5_IT_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse2_L5_IT_Flt1_3UTR.wig\n",
      "Mouse3_L5_IT_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_L5_IT_Flt1_3UTR.wig\n",
      "Mouse4_L5_IT_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse4_L5_IT_Flt1_3UTR.wig\n",
      "ss2_L5_IT_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/ss2_L5_IT_Flt1_3UTR.wig\n",
      "Mouse1_L6_CT_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_L6_CT_Flt1_3UTR.wig\n",
      "Mouse2_L6_CT_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse2_L6_CT_Flt1_3UTR.wig\n",
      "Mouse3_L6_CT_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_L6_CT_Flt1_3UTR.wig\n",
      "Mouse4_L6_CT_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse4_L6_CT_Flt1_3UTR.wig\n",
      "ss2_L6_CT_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/ss2_L6_CT_Flt1_3UTR.wig\n",
      "Mouse1_SMC_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_SMC_Flt1_3UTR.wig\n",
      "Mouse3_SMC_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_SMC_Flt1_3UTR.wig\n",
      "ss2_SMC_CDF /scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/ss2_SMC_Flt1_3UTR.wig\n"
     ]
    }
   ],
   "source": [
    "name = 'Flt1_3UTR'\n",
    "chrom,start,end = 'chr5',147556909,147643009\n",
    "\n",
    "\n",
    "wiggle_header_stem = 'track type=wiggle_0 name={} visibility=full maxHeightPixels=100:50:20 graphType=point color={},{},{}\\nfixedStep chrom={} start={} step=1\\n'\n",
    "\n",
    "\n",
    "ont_palettes = {\n",
    "    'Endo':sns.color_palette('Blues'),\n",
    "    'L5_IT':sns.color_palette('Oranges'),\n",
    "    'L6_CT':sns.color_palette('Greens'),\n",
    "    'SMC':sns.color_palette('Reds'),\n",
    "}\n",
    "\n",
    "for ont,palette in ont_palettes.items():\n",
    "    tenX_stem = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/*{}.bam'.format(ont)\n",
    "    bam_paths = sorted(glob.glob(tenX_stem))\n",
    "    ss2_stem = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/*{}.bam'.format(ont)\n",
    "    bam_paths += sorted(glob.glob(ss2_stem))\n",
    "    \n",
    "    for i,bam_path in enumerate(bam_paths):\n",
    "        color = palette[i]\n",
    "        r,g,b = [int(255*c) for c in color]\n",
    "\n",
    "        with pysam.AlignmentFile(bam_path) as bam:\n",
    "            sample_name = os.path.basename(bam_path).split('.')[0]+'_CDF'\n",
    "            wiggle_header = wiggle_header_stem.format(\n",
    "                sample_name,\n",
    "                r,g,b,\n",
    "                chrom,start,end-start+1,\n",
    "            )\n",
    "            \n",
    "            wig_path = bam_path.replace('.bam','_{}.wig'.format(name))\n",
    "            print(sample_name,wig_path)\n",
    "            with open(wig_path,'w') as wig:\n",
    "                wig.write(wiggle_header)\n",
    "                base_counts = {pos:0 for pos in range(start,end+1)}\n",
    "                for r in bam.fetch(chrom,start,end):\n",
    "                    if r.pos not in base_counts:\n",
    "                        continue\n",
    "                    base_counts[r.pos] += 1\n",
    "\n",
    "                cum_counts = np.cumsum(list(base_counts.values()))\n",
    "                cum_counts = cum_counts/max(cum_counts)\n",
    "                \n",
    "                wig.write('\\n'.join(str(c) for c in cum_counts))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_L5_IT.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse2_L5_IT.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse4_L5_IT.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse4_Endo.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_L5_IT.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_Endo.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_Endo.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse2_Endo.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/ss2_L5_IT.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/ss2_Endo.bam']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "bam_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "L5_IT /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_L5_IT.bam\n",
      "Endo /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_Endo.bam\n",
      "got here\n",
      "L5_IT /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse2_L5_IT.bam\n",
      "Endo /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse2_Endo.bam\n",
      "got here\n",
      "L5_IT /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse4_L5_IT.bam\n",
      "Endo /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse4_Endo.bam\n",
      "got here\n",
      "L5_IT /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_L5_IT.bam\n",
      "Endo /scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_Endo.bam\n",
      "got here\n",
      "L5_IT /scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/ss2_L5_IT.bam\n",
      "Endo /scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/ss2_Endo.bam\n",
      "got here\n"
     ]
    }
   ],
   "source": [
    "#Subtracting values from paired mice\n",
    "\n",
    "name = 'Flt1_3UTR'\n",
    "chrom,start,end = 'chr5',147556909,147643009\n",
    "\n",
    "\n",
    "wiggle_header_stem = 'track type=wiggle_0 name={} visibility=full maxHeightPixels=100:50:20 graphType=point color={},{},{}\\nfixedStep chrom={} start={} step=1\\n'\n",
    "\n",
    "\n",
    "ont_palettes = {\n",
    "    'Endo':sns.color_palette('Blues'),\n",
    "    'L5_IT':sns.color_palette('Oranges'),\n",
    "}\n",
    "\n",
    "tenX_stem = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/*.bam'\n",
    "bam_paths = [p for p in glob.glob(tenX_stem) if any(k in p for k in ont_palettes)]\n",
    "\n",
    "ss2_stem = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/*.bam'\n",
    "bam_paths += [p for p in glob.glob(ss2_stem) if any(k in p for k in ont_palettes)]\n",
    "\n",
    "bams_per_mouse = {}\n",
    "\n",
    "for p in bam_paths:\n",
    "    mouse = os.path.basename(p).split('_')[0]\n",
    "    ont = 'Endo' if 'Endo' in p else 'L5_IT'\n",
    "    tech = 'ss2' if 'ss2' in p else 'tenX'\n",
    "    if (mouse,tech) not in bams_per_mouse:\n",
    "        bams_per_mouse[(mouse,tech)] = {ont:p}\n",
    "    else:\n",
    "        bams_per_mouse[(mouse,tech)][ont] = p\n",
    "        \n",
    "cumsum_counts = {m:{} for m in bams_per_mouse}\n",
    "\n",
    "for (mouse,tech),paths_dict in bams_per_mouse.items():\n",
    "\n",
    "    \n",
    "    for ont,bam_path in paths_dict.items():\n",
    "        print(ont,bam_path)\n",
    "        with pysam.AlignmentFile(bam_path) as bam:\n",
    "            wiggle_header = wiggle_header_stem.format(\n",
    "                mouse,\n",
    "                0,0,0,\n",
    "                chrom,start,end-start+1,\n",
    "            )\n",
    "\n",
    "            base_counts = {pos:0 for pos in range(start,end+1)}\n",
    "            for r in bam.fetch(chrom,start,end):\n",
    "                if r.pos not in base_counts:\n",
    "                    continue\n",
    "                base_counts[r.pos] += 1\n",
    "\n",
    "            cum_counts = np.cumsum(list(base_counts.values()))\n",
    "            cum_counts = cum_counts/max(cum_counts)\n",
    "            cumsum_counts[(mouse,tech)][ont] = cum_counts\n",
    "            \n",
    "    print('got here')\n",
    "    div_cumsum_counts = cumsum_counts[(mouse,tech)]['Endo']-cumsum_counts[(mouse,tech)]['L5_IT']\n",
    "\n",
    "    wig_path = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Flt1_{}_{}_subtract.wig'.format(mouse,tech)\n",
    "    with open(wig_path,'w') as wig:\n",
    "        wig.write(wiggle_header)\n",
    "        wig.write('\\n'.join(str(c) for c in div_cumsum_counts))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('Mouse1',\n",
       "  'tenX'): {'Endo': '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_Endo.bam'},\n",
       " ('Mouse2',\n",
       "  'tenX'): {'Endo': '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse2_Endo.bam'},\n",
       " ('Mouse3',\n",
       "  'tenX'): {'Endo': '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_Endo.bam'},\n",
       " ('Mouse4',\n",
       "  'tenX'): {'Endo': '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse4_Endo.bam'},\n",
       " ('ss2',\n",
       "  'ss2'): {'Endo': '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/ss2_Endo.bam'}}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bams_per_mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_L5_IT.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse2_L5_IT.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse4_L5_IT.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse4_Endo.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_L5_IT.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_Endo.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_Endo.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse2_Endo.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/ss2_L5_IT.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/ss2_Endo.bam']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bam_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('Mouse1', 'tenX'): {'Endo': array([0., 0., 0., ..., 1., 1., 1.])},\n",
       " ('Mouse2', 'tenX'): {},\n",
       " ('Mouse3', 'tenX'): {},\n",
       " ('Mouse4', 'tenX'): {},\n",
       " ('ss2', 'ss2'): {}}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cumsum_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_SMC.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_SMC.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/ss2_SMC.bam']"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bam_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "name = 'ubash3b_3UTR'\n",
    "chrom,start,end = 'chr9',41008509,41017570\n",
    "\n",
    "\n",
    "wiggle_header_stem = 'track type=wiggle_0 name={} visibility=full maxHeightPixels=100:50:20 graphType=point color={},{},{}\\nfixedStep chrom={} start={} step=1\\n'\n",
    "\n",
    "\n",
    "ont_palettes = {\n",
    "    'Sst':sns.color_palette('Blues'),\n",
    "    'L6_IT':sns.color_palette('Oranges'),\n",
    "    'Oligo':sns.color_palette('Greens'),\n",
    "}\n",
    "\n",
    "for ont,palette in ont_palettes.items():\n",
    "    stem = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/*{}.bam'.format(ont)\n",
    "    bam_paths = sorted(glob.glob(stem))\n",
    "    \n",
    "    for i,bam_path in enumerate(bam_paths):\n",
    "        color = palette[i]\n",
    "        r,g,b = [int(255*c) for c in color]\n",
    "\n",
    "        with pysam.AlignmentFile(bam_path) as bam:\n",
    "            sample_name = os.path.basename(bam_path).split('.')[0]+'_CDF'\n",
    "            wiggle_header = wiggle_header_stem.format(\n",
    "                sample_name,\n",
    "                r,g,b,\n",
    "                chrom,start,end-start+1,\n",
    "            )\n",
    "            wig_path = bam_path.replace('.bam','_{}.wig'.format(name))\n",
    "            with open(wig_path,'w') as wig:\n",
    "                wig.write(wiggle_header)\n",
    "                base_counts = {pos:0 for pos in range(start,end+1)}\n",
    "                for r in bam.fetch(chrom,start,end):\n",
    "                    if r.pos not in base_counts:\n",
    "                        continue\n",
    "                    base_counts[r.pos] += 1\n",
    "\n",
    "                cum_counts = np.cumsum(list(base_counts.values()))\n",
    "                cum_counts = cum_counts/max(cum_counts)\n",
    "                \n",
    "                wig.write('\\n'.join(str(c) for c in cum_counts))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "name = 'itgb8_5UTR'\n",
    "chrom,start,end = 'chr12',119158022,119164343\n",
    "\n",
    "\n",
    "wiggle_header_stem = 'track type=wiggle_0 name={} visibility=full maxHeightPixels=100:50:20 graphType=point color={},{},{}\\nfixedStep chrom={} start={} step=1\\n'\n",
    "\n",
    "\n",
    "ont_palettes = {\n",
    "    'Oligo':sns.color_palette('Blues'),\n",
    "    'L6_CT':sns.color_palette('Reds'),\n",
    "}\n",
    "\n",
    "for ont,palette in ont_palettes.items():\n",
    "    stem = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/*{}.bam'.format(ont)\n",
    "    bam_paths = sorted(glob.glob(stem))\n",
    "    \n",
    "    for i,bam_path in enumerate(bam_paths):\n",
    "        color = palette[i]\n",
    "        r,g,b = [int(255*c) for c in color]\n",
    "\n",
    "        with pysam.AlignmentFile(bam_path) as bam:\n",
    "            sample_name = os.path.basename(bam_path).split('.')[0]+'_CDF'\n",
    "            wiggle_header = wiggle_header_stem.format(\n",
    "                sample_name,\n",
    "                r,g,b,\n",
    "                chrom,start,end-start+1,\n",
    "            )\n",
    "            wig_path = bam_path.replace('.bam','_{}.wig'.format(name))\n",
    "            with open(wig_path,'w') as wig:\n",
    "                wig.write(wiggle_header)\n",
    "                base_counts = {pos:0 for pos in range(start,end+1)}\n",
    "                for r in bam.fetch(chrom,start,end):\n",
    "                    if r.pos not in base_counts:\n",
    "                        continue\n",
    "                    base_counts[r.pos] += 1\n",
    "\n",
    "                cum_counts = np.cumsum(list(base_counts.values()))\n",
    "                cum_counts = cum_counts/max(cum_counts)\n",
    "                \n",
    "                wig.write('\\n'.join(str(c) for c in cum_counts))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Subtracting values from paired mice\n",
    "\n",
    "name = 'itgb8_5UTR'\n",
    "chrom,start,end = 'chr12',119158022,119164343\n",
    "\n",
    "\n",
    "wiggle_header_stem = 'track type=wiggle_0 name={} visibility=full maxHeightPixels=100:50:20 graphType=point color={},{},{}\\nfixedStep chrom={} start={} step=1\\n'\n",
    "\n",
    "\n",
    "ont_palettes = {\n",
    "    'Oligo':sns.color_palette('Blues'),\n",
    "    'L6_CT':sns.color_palette('Reds'),\n",
    "}\n",
    "\n",
    "\n",
    "stem = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/*.bam'\n",
    "bam_paths = [p for p in glob.glob(stem) if any(k in p for k in ont_palettes)]\n",
    "bams_per_mouse = {}\n",
    "\n",
    "for p in bam_paths:\n",
    "    mouse = os.path.basename(p).split('_')[0]\n",
    "    ont = 'Oligo' if 'Oligo' in p else 'L6_CT'\n",
    "    if mouse not in bams_per_mouse:\n",
    "        bams_per_mouse[mouse] = {ont:p}\n",
    "    else:\n",
    "        bams_per_mouse[mouse][ont] = p\n",
    "        \n",
    "cumsum_counts = {m:{} for m in bams_per_mouse}\n",
    "\n",
    "for mouse,paths_dict in bams_per_mouse.items():\n",
    "\n",
    "    \n",
    "    for ont,bam_path in paths_dict.items():\n",
    "        with pysam.AlignmentFile(bam_path) as bam:\n",
    "            wiggle_header = wiggle_header_stem.format(\n",
    "                mouse,\n",
    "                0,0,0,\n",
    "                chrom,start,end-start+1,\n",
    "            )\n",
    "\n",
    "            base_counts = {pos:0 for pos in range(start,end+1)}\n",
    "            for r in bam.fetch(chrom,start,end):\n",
    "                if r.pos not in base_counts:\n",
    "                    continue\n",
    "                base_counts[r.pos] += 1\n",
    "\n",
    "            cum_counts = np.cumsum(list(base_counts.values()))\n",
    "            cum_counts = cum_counts/max(cum_counts)\n",
    "            cumsum_counts[mouse][ont] = cum_counts\n",
    "            \n",
    "    div_cumsum_counts = cumsum_counts[mouse]['Oligo']-cumsum_counts[mouse]['L6_CT']\n",
    "\n",
    "    wig_path = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/itgb8_{}_subtract.wig'.format(mouse)\n",
    "    with open(wig_path,'w') as wig:\n",
    "        wig.write(wiggle_header)\n",
    "        wig.write('\\n'.join(str(c) for c in div_cumsum_counts))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Mouse1': {'Oligo': '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_Oligo.bam',\n",
       "  'Sst': '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_Sst.bam'},\n",
       " 'Mouse2': {'Oligo': '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse2_Oligo.bam',\n",
       "  'Sst': '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse2_Sst.bam'},\n",
       " 'Mouse3': {'Oligo': '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_Oligo.bam',\n",
       "  'Sst': '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_Sst.bam'},\n",
       " 'Mouse4': {'Oligo': '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse4_Oligo.bam',\n",
       "  'Sst': '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse4_Sst.bam'}}"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Subtracting values from paired mice\n",
    "\n",
    "name = 'ubash3b_3UTR'\n",
    "chrom,start,end = 'chr9',41008509,41017570\n",
    "\n",
    "\n",
    "wiggle_header_stem = 'track type=wiggle_0 name={} visibility=full maxHeightPixels=100:50:20 graphType=point color={},{},{}\\nfixedStep chrom={} start={} step=1\\n'\n",
    "\n",
    "\n",
    "ont_palettes = {\n",
    "    'Sst':sns.color_palette('Blues'),\n",
    "    'Oligo':sns.color_palette('Oranges'),\n",
    "}\n",
    "\n",
    "stem = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/*.bam'\n",
    "bam_paths = [p for p in glob.glob(stem) if any(k in p for k in ont_palettes)]\n",
    "bams_per_mouse = {}\n",
    "\n",
    "for p in bam_paths:\n",
    "    mouse = os.path.basename(p).split('_')[0]\n",
    "    ont = 'Sst' if 'Sst' in p else 'Oligo'\n",
    "    if mouse not in bams_per_mouse:\n",
    "        bams_per_mouse[mouse] = {ont:p}\n",
    "    else:\n",
    "        bams_per_mouse[mouse][ont] = p\n",
    "        \n",
    "cumsum_counts = {m:{} for m in bams_per_mouse}\n",
    "\n",
    "for mouse,paths_dict in bams_per_mouse.items():\n",
    "\n",
    "    \n",
    "    for ont,bam_path in paths_dict.items():\n",
    "        with pysam.AlignmentFile(bam_path) as bam:\n",
    "            wiggle_header = wiggle_header_stem.format(\n",
    "                mouse,\n",
    "                0,0,0,\n",
    "                chrom,start,end-start+1,\n",
    "            )\n",
    "\n",
    "            base_counts = {pos:0 for pos in range(start,end+1)}\n",
    "            for r in bam.fetch(chrom,start,end):\n",
    "                if r.pos not in base_counts:\n",
    "                    continue\n",
    "                base_counts[r.pos] += 1\n",
    "\n",
    "            cum_counts = np.cumsum(list(base_counts.values()))\n",
    "            cum_counts = cum_counts/max(cum_counts)\n",
    "            cumsum_counts[mouse][ont] = cum_counts\n",
    "            \n",
    "    div_cumsum_counts = cumsum_counts[mouse]['Oligo']-cumsum_counts[mouse]['Sst']\n",
    "\n",
    "    wig_path = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/ubash3b_{}_subtract.wig'.format(mouse)\n",
    "    with open(wig_path,'w') as wig:\n",
    "        wig.write(wiggle_header)\n",
    "        wig.write('\\n'.join(str(c) for c in div_cumsum_counts))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Mouse1': {'Oligo': array([0., 0., 0., ..., 1., 1., 1.]),\n",
       "  'Sst': array([0., 0., 0., ..., 1., 1., 1.])},\n",
       " 'Mouse2': {'Oligo': array([0., 0., 0., ..., 1., 1., 1.]),\n",
       "  'Sst': array([0., 0., 0., ..., 1., 1., 1.])},\n",
       " 'Mouse3': {'Oligo': array([0., 0., 0., ..., 1., 1., 1.]),\n",
       "  'Sst': array([0., 0., 0., ..., 1., 1., 1.])},\n",
       " 'Mouse4': {'Oligo': array([0., 0., 0., ..., 1., 1., 1.]),\n",
       "  'Sst': array([0., 0., 0., ..., 1., 1., 1.])}}"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cumsum_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mouse1 KstestResult(statistic=0.1965349812403443, pvalue=1.994987199237887e-153)\n",
      "Mouse4 KstestResult(statistic=0.23317148532332818, pvalue=2.2786690426516e-216)\n",
      "Mouse2 KstestResult(statistic=0.20459059810196425, pvalue=2.563058849593599e-166)\n",
      "Mouse3 KstestResult(statistic=0.2700286912381373, pvalue=5.750717485230399e-291)\n"
     ]
    }
   ],
   "source": [
    "import scipy\n",
    "\n",
    "for mouse in cumsum_counts:\n",
    "    v1 = cumsum_counts[mouse]['Sst']\n",
    "    v2 = cumsum_counts[mouse]['Oligo']\n",
    "    print(mouse,scipy.stats.ks_2samp(v1,v2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse1_Oligo.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse2_Oligo.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse3_Oligo.bam',\n",
       " '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/Mouse4_Oligo.bam']"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bam_paths"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Counting reads from different mice/onts for regions of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>num_reads</th>\n",
       "      <th>mouse</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>0</td>\n",
       "      <td>Mouse4</td>\n",
       "      <td>Macrophage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>1</td>\n",
       "      <td>Mouse4</td>\n",
       "      <td>Sncg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>0</td>\n",
       "      <td>Mouse1</td>\n",
       "      <td>Oligo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>1</td>\n",
       "      <td>Mouse1</td>\n",
       "      <td>Macrophage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>166</td>\n",
       "      <td>Mouse2</td>\n",
       "      <td>L6_CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>14</td>\n",
       "      <td>Mouse4</td>\n",
       "      <td>Sst</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>21</td>\n",
       "      <td>Mouse2</td>\n",
       "      <td>Lamp5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>0</td>\n",
       "      <td>Mouse3</td>\n",
       "      <td>Astro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>130</td>\n",
       "      <td>Mouse3</td>\n",
       "      <td>Pvalb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>97</td>\n",
       "      <td>Mouse2</td>\n",
       "      <td>L6_IT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    name  num_reads   mouse         ont\n",
       "0   Flt1          0  Mouse4  Macrophage\n",
       "1   Flt1          1  Mouse4        Sncg\n",
       "2   Flt1          0  Mouse1       Oligo\n",
       "3   Flt1          1  Mouse1  Macrophage\n",
       "4   Flt1        166  Mouse2       L6_CT\n",
       "..   ...        ...     ...         ...\n",
       "73  Flt1         14  Mouse4         Sst\n",
       "74  Flt1         21  Mouse2       Lamp5\n",
       "75  Flt1          0  Mouse3       Astro\n",
       "76  Flt1        130  Mouse3       Pvalb\n",
       "77  Flt1         97  Mouse2       L6_IT\n",
       "\n",
       "[78 rows x 4 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions = {\n",
    "    'Flt1':('chr5',147520502,147767113),\n",
    "}\n",
    "\n",
    "\n",
    "stem = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/10X_mapping/merged_by_mouse_celltype/*.bam'\n",
    "bam_paths = glob.glob(stem)\n",
    "\n",
    "data = {\n",
    "    'name':[],\n",
    "    'num_reads':[],\n",
    "    'mouse':[],\n",
    "    'ont':[],\n",
    "}\n",
    "\n",
    "for bam_path in bam_paths:\n",
    "    bam_base = os.path.basename(bam_path).replace('.bam','')\n",
    "    mouse = bam_base.split('_')[0]\n",
    "    ont = '_'.join(bam_base.split('_')[1:])\n",
    "    bam = pysam.AlignmentFile(bam_path)\n",
    "    \n",
    "    for name,region in regions.items():\n",
    "        rs = bam.fetch(*region)\n",
    "        num_reads = sum(1 for _ in rs)\n",
    "        \n",
    "        data['name'].append(name)\n",
    "        data['num_reads'].append(num_reads)\n",
    "        data['mouse'].append(mouse)\n",
    "        data['ont'].append(ont)\n",
    "        \n",
    "region_counts_df = pd.DataFrame(data)\n",
    "region_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>num_reads</th>\n",
       "      <th>mouse</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>0</td>\n",
       "      <td>ss2</td>\n",
       "      <td>L5_ET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>0</td>\n",
       "      <td>ss2</td>\n",
       "      <td>VLMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>0</td>\n",
       "      <td>ss2</td>\n",
       "      <td>Astro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>1</td>\n",
       "      <td>ss2</td>\n",
       "      <td>L6_IT_Car3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>5</td>\n",
       "      <td>ss2</td>\n",
       "      <td>Sncg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>25</td>\n",
       "      <td>ss2</td>\n",
       "      <td>Vip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>32</td>\n",
       "      <td>ss2</td>\n",
       "      <td>Pvalb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>35</td>\n",
       "      <td>ss2</td>\n",
       "      <td>L56_NP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>64</td>\n",
       "      <td>ss2</td>\n",
       "      <td>L23_IT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>71</td>\n",
       "      <td>ss2</td>\n",
       "      <td>Lamp5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>128</td>\n",
       "      <td>ss2</td>\n",
       "      <td>Sst</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>162</td>\n",
       "      <td>ss2</td>\n",
       "      <td>L6_IT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>329</td>\n",
       "      <td>ss2</td>\n",
       "      <td>L5_IT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>416</td>\n",
       "      <td>ss2</td>\n",
       "      <td>L6_CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>679</td>\n",
       "      <td>ss2</td>\n",
       "      <td>L6b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>1036</td>\n",
       "      <td>ss2</td>\n",
       "      <td>SMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>14837</td>\n",
       "      <td>ss2</td>\n",
       "      <td>Endo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    name  num_reads mouse         ont\n",
       "8   Flt1          0   ss2       L5_ET\n",
       "1   Flt1          0   ss2        VLMC\n",
       "11  Flt1          0   ss2       Astro\n",
       "3   Flt1          1   ss2  L6_IT_Car3\n",
       "13  Flt1          5   ss2        Sncg\n",
       "10  Flt1         25   ss2         Vip\n",
       "15  Flt1         32   ss2       Pvalb\n",
       "0   Flt1         35   ss2      L56_NP\n",
       "14  Flt1         64   ss2      L23_IT\n",
       "12  Flt1         71   ss2       Lamp5\n",
       "9   Flt1        128   ss2         Sst\n",
       "4   Flt1        162   ss2       L6_IT\n",
       "2   Flt1        329   ss2       L5_IT\n",
       "5   Flt1        416   ss2       L6_CT\n",
       "7   Flt1        679   ss2         L6b\n",
       "6   Flt1       1036   ss2         SMC\n",
       "16  Flt1      14837   ss2        Endo"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions = {\n",
    "    'Flt1':('chr5',147520502,147767113),\n",
    "}\n",
    "\n",
    "\n",
    "stem = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype/*.bam'\n",
    "bam_paths = glob.glob(stem)\n",
    "\n",
    "data = {\n",
    "    'name':[],\n",
    "    'num_reads':[],\n",
    "    'mouse':[],\n",
    "    'ont':[],\n",
    "}\n",
    "\n",
    "for bam_path in bam_paths:\n",
    "    bam_base = os.path.basename(bam_path).replace('.bam','')\n",
    "    mouse = bam_base.split('_')[0]\n",
    "    ont = '_'.join(bam_base.split('_')[1:])\n",
    "    bam = pysam.AlignmentFile(bam_path)\n",
    "    \n",
    "    for name,region in regions.items():\n",
    "        rs = bam.fetch(*region)\n",
    "        num_reads = sum(1 for _ in rs)\n",
    "        \n",
    "        data['name'].append(name)\n",
    "        data['num_reads'].append(num_reads)\n",
    "        data['mouse'].append(mouse)\n",
    "        data['ont'].append(ont)\n",
    "        \n",
    "region_counts_df = pd.DataFrame(data)\n",
    "region_counts_df.sort_values('num_reads')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>num_reads</th>\n",
       "      <th>mouse</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>102</td>\n",
       "      <td>Mouse2</td>\n",
       "      <td>VLMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>106</td>\n",
       "      <td>Mouse2</td>\n",
       "      <td>L23_IT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>121</td>\n",
       "      <td>Mouse4</td>\n",
       "      <td>L23_IT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>130</td>\n",
       "      <td>Mouse3</td>\n",
       "      <td>Pvalb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>131</td>\n",
       "      <td>Mouse3</td>\n",
       "      <td>Lamp5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>166</td>\n",
       "      <td>Mouse2</td>\n",
       "      <td>L6_CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>189</td>\n",
       "      <td>Mouse4</td>\n",
       "      <td>L6_CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>200</td>\n",
       "      <td>Mouse1</td>\n",
       "      <td>L23_IT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>290</td>\n",
       "      <td>Mouse4</td>\n",
       "      <td>L5_IT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>406</td>\n",
       "      <td>Mouse1</td>\n",
       "      <td>L5_IT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>898</td>\n",
       "      <td>Mouse2</td>\n",
       "      <td>L5_IT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>3298</td>\n",
       "      <td>Mouse4</td>\n",
       "      <td>Endo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>5000</td>\n",
       "      <td>Mouse3</td>\n",
       "      <td>Endo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>5262</td>\n",
       "      <td>Mouse2</td>\n",
       "      <td>Endo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Flt1</td>\n",
       "      <td>10732</td>\n",
       "      <td>Mouse1</td>\n",
       "      <td>Endo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    name  num_reads   mouse     ont\n",
       "22  Flt1        102  Mouse2    VLMC\n",
       "53  Flt1        106  Mouse2  L23_IT\n",
       "31  Flt1        121  Mouse4  L23_IT\n",
       "76  Flt1        130  Mouse3   Pvalb\n",
       "50  Flt1        131  Mouse3   Lamp5\n",
       "4   Flt1        166  Mouse2   L6_CT\n",
       "44  Flt1        189  Mouse4   L6_CT\n",
       "34  Flt1        200  Mouse1  L23_IT\n",
       "36  Flt1        290  Mouse4   L5_IT\n",
       "20  Flt1        406  Mouse1   L5_IT\n",
       "26  Flt1        898  Mouse2   L5_IT\n",
       "41  Flt1       3298  Mouse4    Endo\n",
       "55  Flt1       5000  Mouse3    Endo\n",
       "63  Flt1       5262  Mouse2    Endo\n",
       "48  Flt1      10732  Mouse1    Endo"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_counts_df[region_counts_df['num_reads'].ge(100)].sort_values('num_reads')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Split SS2 main BAM by annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-54-f5e100477df1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     78\u001b[0m     \u001b[0mann\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbc_to_ann\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     79\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mann\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mpass_filter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 80\u001b[0;31m         \u001b[0mout_bams\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mann\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     81\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     82\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "import glob\n",
    "import sys\n",
    "import os\n",
    "\n",
    "\n",
    "#Sam flag meanings from https://broadinstitute.github.io/picard/explain-flags.html\n",
    "bit_meanings = {\n",
    "    'read paired':                               0b000000000001,\n",
    "    'read mapped in proper pair':                0b000000000010,\n",
    "    'read unmapped':                             0b000000000100,\n",
    "    'mate unmapped':                             0b000000001000,\n",
    "    'read reverse strand':                       0b000000010000,\n",
    "    'mate reverse strand':                       0b000000100000,\n",
    "    'first in pair':                             0b000001000000,\n",
    "    'second in pair':                            0b000010000000,\n",
    "    'not primary alignment':                     0b000100000000,\n",
    "    'read fails platform/vendor quality checks': 0b001000000000,\n",
    "    'read is PCR or optical duplicate':          0b010000000000,\n",
    "    'supplementary alignment':                   0b100000000000,\n",
    "}\n",
    "\n",
    "#If any of these bits are set, reject the read\n",
    "#can test if a flag is rejectable if flag&reject_flags != 0\n",
    "reject_flags = (\n",
    "    bit_meanings['read unmapped'] |\n",
    "    bit_meanings['mate unmapped'] |\n",
    "    bit_meanings['second in pair'] |\n",
    "    bit_meanings['not primary alignment'] |\n",
    "    bit_meanings['read fails platform/vendor quality checks'] |\n",
    "    bit_meanings['supplementary alignment'] |\n",
    "    bit_meanings['read is PCR or optical duplicate']\n",
    ")\n",
    "\n",
    "def pass_filter(read):\n",
    "    cigar = read.cigar\n",
    "    mq = read.mapping_quality\n",
    "    read_length = read.query_length\n",
    "    rejectable_flag = read.flag & reject_flags\n",
    "    if cigar == [(0, read_length)] and mq == 255 and not rejectable_flag:\n",
    "        return True\n",
    "\n",
    "\n",
    "bam_path = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/SS2_merged_filtered.bam'\n",
    "meta_path = '/oak/stanford/groups/horence/rob/readzs_fork/MOp_SS2_metadata.tsv'\n",
    "\n",
    "meta_df = pd.read_csv(meta_path,sep='\\t')\n",
    "meta_df['subclass_label'] = meta_df['subclass_label'].str.replace('/','')\n",
    "meta_df['subclass_label'] = meta_df['subclass_label'].str.replace(' ','_')\n",
    "meta_df['cell_id'] = meta_df['cell_id'].str.replace('SS2_','')\n",
    "\n",
    "bc_to_ann = dict(meta_df[['cell_id','subclass_label']].values)\n",
    "\n",
    "\n",
    "#Get file handle to bam to split\n",
    "in_bam = pysam.AlignmentFile(bam_path, 'rb')\n",
    "\n",
    "\n",
    "#Prepare output bams for each possible cell-type\n",
    "out_stem = '/scratch/groups/horence/rob/data/MERFISH_scRNAseq/ss2_post_picard_bam_tag_filtering/split_by_celltype'\n",
    "out_bams = {}\n",
    "out_paths = []\n",
    "for ann in meta_df['subclass_label'].unique():\n",
    "    out_path = os.path.join(out_stem,ann+'.bam')\n",
    "    out_bam = pysam.AlignmentFile(out_path,'wb',template=in_bam)\n",
    "    out_bams[ann] = out_bam\n",
    "    out_paths.append(out_path)\n",
    "\n",
    "#Iterate through the input bam\n",
    "for i,r in enumerate(in_bam.fetch()):\n",
    "    try:\n",
    "        bc = r.get_tag('CB')\n",
    "    except KeyError:\n",
    "        continue\n",
    "\n",
    "    ann = bc_to_ann.get(bc)\n",
    "    if ann and pass_filter(r):\n",
    "        out_bams[ann].write(r)\n",
    "\n",
    "\n",
    "#Close the output bams\n",
    "for ann,bam_f in out_bams.items():\n",
    "    bam_f.close()\n",
    "\n",
    "\n",
    "#Index the bams\n",
    "for out_path in out_paths:\n",
    "    pysam.index(out_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
